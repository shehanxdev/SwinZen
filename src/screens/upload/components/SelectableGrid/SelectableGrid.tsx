import React, { useState } from 'react';
import { Pressable, View } from 'react-native';

import { Text } from '@sz/components';
import { tw } from '@sz/config';
import { Color, TextVariant } from '@sz/constants';

interface SelectableGridProps {
  testID?: string;
  options: { label: string; value: string }[];
  onChange: (value: string) => void;
}

export function SelectableGrid({ options, onChange, testID }: SelectableGridProps) {
  const [selectedCellIndex, setSelectedCellIndex] = useState(0);

  const onPress = (index: number, value: string) => {
    setSelectedCellIndex(index);
    onChange(value);
  };

  return (
    <View style={tw`flex-row flex-wrap justify-between gap-2`} testID={testID}>
      {options.map((option, index) => {
        return (
          <Pressable
            testID="grid-option"
            key={index}
            onPress={() => onPress(index, option.value)}
            style={tw`grow w-27.5 h-9  rounded-2 justify-center  ${
              index === selectedCellIndex
                ? `bg-[${Color.Tertiary.Sz900}] text-[${Color.Primary.Sz900}]`
                : `bg-[${Color.Neutral.Black}] text-[${Color.Neutral.Sz700}]`
            }`}>
            <Text
              variant={TextVariant.Body2SemiBold}
              color={index === selectedCellIndex ? Color.Primary.Sz900 : Color.Neutral.Sz700}>
              {option.label}
            </Text>
          </Pressable>
        );
      })}
    </View>
  );
}
