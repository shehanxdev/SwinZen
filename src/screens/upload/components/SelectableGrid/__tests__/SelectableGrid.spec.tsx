import { fireEvent, render } from '@testing-library/react-native';
import * as React from 'react';

import { SelectableGrid } from '../SelectableGrid';

describe('SelectableGrid Component', () => {
  const testID = 'SelectableGridTestID';
  const onChangeMock = jest.fn();
  const options = [
    { label: 'Option 1', value: 'option1' },
    { label: 'Option 2', value: 'option2' },
    { label: 'Option 3', value: 'option3' },
  ];

  const getRenderedComponent = () =>
    render(<SelectableGrid options={options} onChange={onChangeMock} testID={testID} />);

  it(`should render SelectableGrid component correctly`, () => {
    const rendered = getRenderedComponent();
    const renderedTree = rendered.toJSON();
    expect(renderedTree).toMatchSnapshot();
  });

  it(`should find the SelectableGrid component via testID`, () => {
    const { getByTestId } = getRenderedComponent();
    const foundCustomHeaderComponent = getByTestId(testID);
    expect(foundCustomHeaderComponent).toBeTruthy();
  });

  it('displays the correct number of options', () => {
    const { getAllByTestId } = getRenderedComponent();
    const optionElements = getAllByTestId('grid-option');
    expect(optionElements.length).toBe(options.length);
  });

  it('calls the onChange function with the selected value', () => {
    const { getByText } = getRenderedComponent();
    const optionElement = getByText('Option 2');

    fireEvent.press(optionElement);

    expect(onChangeMock).toHaveBeenCalledTimes(1);
    expect(onChangeMock).toHaveBeenCalledWith('option2');
  });
});
