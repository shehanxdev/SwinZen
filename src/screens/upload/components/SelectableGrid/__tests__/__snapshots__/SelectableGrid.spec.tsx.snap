// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectableGrid Component should render SelectableGrid component correctly 1`] = `
<View
  style={
    {
      "flexDirection": "row",
      "flexWrap": "wrap",
      "gap": 8,
      "justifyContent": "space-between",
    }
  }
  testID="SelectableGridTestID"
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "backgroundColor": "#A2FD2F",
        "borderRadius": 8,
        "color": "#070707",
        "flexGrow": 1,
        "height": 36,
        "justifyContent": "center",
        "width": 110,
      }
    }
    testID="grid-option"
  >
    <Text
      style={
        [
          {
            "color": "#070707",
            "lineHeight": 24,
          },
          {
            "fontFamily": "SourceSansPro-SemiBold",
            "fontSize": 16,
            "fontWeight": "600",
            "textAlign": "center",
          },
        ]
      }
    >
      Option 1
    </Text>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "backgroundColor": "#000000",
        "borderRadius": 8,
        "color": "#495057",
        "flexGrow": 1,
        "height": 36,
        "justifyContent": "center",
        "width": 110,
      }
    }
    testID="grid-option"
  >
    <Text
      style={
        [
          {
            "color": "#495057",
            "lineHeight": 24,
          },
          {
            "fontFamily": "SourceSansPro-SemiBold",
            "fontSize": 16,
            "fontWeight": "600",
            "textAlign": "center",
          },
        ]
      }
    >
      Option 2
    </Text>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "backgroundColor": "#000000",
        "borderRadius": 8,
        "color": "#495057",
        "flexGrow": 1,
        "height": 36,
        "justifyContent": "center",
        "width": 110,
      }
    }
    testID="grid-option"
  >
    <Text
      style={
        [
          {
            "color": "#495057",
            "lineHeight": 24,
          },
          {
            "fontFamily": "SourceSansPro-SemiBold",
            "fontSize": 16,
            "fontWeight": "600",
            "textAlign": "center",
          },
        ]
      }
    >
      Option 3
    </Text>
  </View>
</View>
`;
